File: typescript/tests/gpu-runtime.test.ts
/**
 * GPU runtime cascade — Phase 2 Task 3.
 * Mirrors python/tests/test_gpu_runtime.py.
 *
 * The TS deviation is the dependency-injection pattern: instead of
 * monkeypatching module attributes, the resolver accepts an Options object
 * exposing `nvmlAvailable`, `getDeviceCount`, and `getCloudEnv` hooks so
 * tests can pass deterministic stubs without import-cycle gymnastics.
 */

import { describe, it, expect, beforeEach, afterEach } from "vitest";
import {
  GpuRuntimeKind,
  resolveGpuRuntime,
  type ResolveGpuRuntimeOptions,
} from "../src/core/gpu-runtime.js";

const SERVERLESS_VARS = [
  "MODAL_TASK_ID",
  "MODAL_IMAGE_ID",
  "RUNPOD_POD_ID",
  "RUNPOD_POD_HOSTNAME",
  "REPLICATE_MODEL",
  "REPLICATE_PREDICTION_ID",
];

function scrubEnv() {
  for (const v of SERVERLESS_VARS) delete process.env[v];
}

function opts(
  nvml: boolean,
  deviceCount: number,
  provider: string | null,
  instanceType?: string,
  region?: string,
): ResolveGpuRuntimeOptions {
  return {
    nvmlAvailable: () => nvml,
    getDeviceCount: () => deviceCount,
    getCloudEnv: () => ({
      provider,
      region: region ?? null,
      source: "env",
      instanceType: instanceType ?? null,
    }),
  };
}

describe("Serverless GPU clouds — env-var detection wins", () => {
  beforeEach(() => scrubEnv());
  afterEach(() => scrubEnv());

  it("MODAL_TASK_ID → MODAL", () => {
    process.env.MODAL_TASK_ID = "task-abc";
    expect(resolveGpuRuntime(opts(true, 1, "modal"))).toBe(
      GpuRuntimeKind.Modal,
    );
  });

  it("RUNPOD_POD_ID → RUNPOD", () => {
    process.env.RUNPOD_POD_ID = "pod-abc";
    expect(resolveGpuRuntime(opts(true, 1, "runpod"))).toBe(
      GpuRuntimeKind.RunPod,
    );
  });

  it("REPLICATE_MODEL → REPLICATE", () => {
    process.env.REPLICATE_MODEL = "owner/model";
    expect(resolveGpuRuntime(opts(true, 1, "replicate"))).toBe(
      GpuRuntimeKind.Replicate,
    );
  });
});

describe("AWS GPU family detection", () => {
  beforeEach(() => scrubEnv());

  const gpuCases: string[] = [
    "p5.48xlarge",
    "p4d.24xlarge",
    "p4de.24xlarge",
    "p5e.48xlarge",
    "p5en.48xlarge",
    "p3.2xlarge",
    "g4dn.xlarge",
    "g4dn.metal",
    "g5.xlarge",
    "g5g.xlarge",
    "g6.xlarge",
    "g6e.xlarge",
  ];

  it.each(gpuCases)("%s → AWS_EC2_GPU", (it_) => {
    expect(resolveGpuRuntime(opts(true, 1, "aws", it_, "us-east-1"))).toBe(
      GpuRuntimeKind.AwsEc2Gpu,
    );
  });

  const nonGpu = ["c7g.xlarge", "t3.medium", "m7i.large"];
  it.each(nonGpu)("%s → NONE (CPU family)", (it_) => {
    expect(resolveGpuRuntime(opts(true, 0, "aws", it_, "us-east-1"))).toBe(
      GpuRuntimeKind.None,
    );
  });
});

describe("GCP GPU family detection", () => {
  beforeEach(() => scrubEnv());

  const bundled = [
    "a2-highgpu-1g",
    "a2-ultragpu-1g",
    "a3-highgpu-8g",
    "a3-edgegpu-8g",
    "g2-standard-4",
  ];
  it.each(bundled)("%s → GCP_GCE_BUNDLED", (it_) => {
    expect(resolveGpuRuntime(opts(true, 1, "gcp", it_, "us-central1"))).toBe(
      GpuRuntimeKind.GcpGceBundled,
    );
  });

  const n1Cases = ["n1-standard-8", "n1-highmem-4"];
  it.each(n1Cases)("%s → GCP_GCE_N1_ATTACHED (Decision #9)", (it_) => {
    expect(resolveGpuRuntime(opts(true, 1, "gcp", it_, "us-central1"))).toBe(
      GpuRuntimeKind.GcpGceN1Attached,
    );
  });

  it("e2-standard-4 (CPU-only) + 0 devices → NONE", () => {
    expect(
      resolveGpuRuntime(opts(true, 0, "gcp", "e2-standard-4", "us-central1")),
    ).toBe(GpuRuntimeKind.None);
  });
});

describe("Azure GPU family detection", () => {
  beforeEach(() => scrubEnv());

  const azureGpu = [
    "Standard_ND96isr_H100_v5",
    "Standard_ND96amsr_A100_v4",
    "Standard_ND96asr_v4",
    "Standard_NC24ads_A100_v4",
    "Standard_NC6s_v3",
  ];
  it.each(azureGpu)("%s → AZURE_VM_GPU", (it_) => {
    expect(resolveGpuRuntime(opts(true, 1, "azure", it_, "eastus"))).toBe(
      GpuRuntimeKind.AzureVmGpu,
    );
  });

  const azureVgpu = [
    "Standard_NV6ads_A10_v5",
    "Standard_NV12ads_A10_v5",
    "Standard_NV36ads_A10_v5",
    "Standard_NV72ads_A10_v5",
  ];
  it.each(azureVgpu)("%s → AZURE_VM_VGPU (Decision #10)", (it_) => {
    expect(resolveGpuRuntime(opts(true, 1, "azure", it_, "eastus"))).toBe(
      GpuRuntimeKind.AzureVmVgpu,
    );
  });

  const nonGpu = ["Standard_D2s_v3", "Standard_B2ms"];
  it.each(nonGpu)("%s → NONE", (it_) => {
    expect(resolveGpuRuntime(opts(true, 0, "azure", it_, "eastus"))).toBe(
      GpuRuntimeKind.None,
    );
  });
});

describe("Reserved-GPU providers (Lambda Labs / CoreWeave)", () => {
  beforeEach(() => scrubEnv());

  it("lambda_labs via cloud_detect → LAMBDA_LABS", () => {
    expect(resolveGpuRuntime(opts(true, 8, "lambda_labs"))).toBe(
      GpuRuntimeKind.LambdaLabs,
    );
  });

  it("coreweave via cloud_detect → COREWEAVE", () => {
    expect(resolveGpuRuntime(opts(true, 8, "coreweave"))).toBe(
      GpuRuntimeKind.CoreWeave,
    );
  });
});

describe("No NVML / no GPU → NONE", () => {
  beforeEach(() => scrubEnv());

  it("NVML unavailable → NONE (even on a Modal task)", () => {
    process.env.MODAL_TASK_ID = "x";
    expect(resolveGpuRuntime(opts(false, 0, "modal"))).toBe(
      GpuRuntimeKind.None,
    );
  });

  it("NVML reports 0 devices → NONE (even on p5)", () => {
    expect(
      resolveGpuRuntime(opts(true, 0, "aws", "p5.48xlarge", "us-east-1")),
    ).toBe(GpuRuntimeKind.None);
  });

  it("no cloud provider resolved + no env vars → NONE", () => {
    expect(resolveGpuRuntime(opts(true, 0, null))).toBe(GpuRuntimeKind.None);
  });
});

File: go/core/gpu_runtime_test.go
// Task 3 — GPU runtime resolver tests. Mirrors python commit 9bcb0c9.

package core

import (
	"os"
	"testing"

	"github.com/DexwoxBusiness/dexcost-go/cloud"
)

// withNVMLMock sets up a mock NVML backend with N devices and restores
// state at the end of the test.
func withNVMLMock(t *testing.T, devices int) {
	t.Helper()
	resetNVMLForTests()
	mock := &MockNVMLBackend{
		Available:    true,
		DeviceCount:  devices,
		ProductNames: map[int]string{0: "NVIDIA H100 80GB HBM3"},
	}
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)
}

func TestGpuRuntimeKindStringValuesMatchPython(t *testing.T) {
	cases := []struct {
		k    GpuRuntimeKind
		want string
	}{
		{GpuRuntimeModal, "modal"},
		{GpuRuntimeRunpod, "runpod"},
		{GpuRuntimeReplicate, "replicate"},
		{GpuRuntimeLambdaLabs, "lambda_labs"},
		{GpuRuntimeCoreweave, "coreweave"},
		{GpuRuntimeAWSEC2GPU, "aws_ec2_gpu"},
		{GpuRuntimeGCPGCEBundled, "gcp_gce_bundled"},
		{GpuRuntimeGCPGCEN1Attached, "gcp_gce_n1_attached"},
		{GpuRuntimeAzureVMGPU, "azure_vm_gpu"},
		{GpuRuntimeAzureVMVGPU, "azure_vm_vgpu"},
		{GpuRuntimeNone, "none"},
	}
	for _, c := range cases {
		if string(c.k) != c.want {
			t.Errorf("%v = %q; want %q (cross-SDK parity)", c.k, c.k, c.want)
		}
	}
}

func TestResolveGpuRuntimeReturnsNoneWhenNVMLUnavailable(t *testing.T) {
	resetNVMLForTests() // default noop backend
	t.Cleanup(resetNVMLForTests)
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "aws", Region: "us-east-1", InstanceType: "p4d.24xlarge"})
	t.Cleanup(cloud.ResetForTests)
	t.Setenv("MODAL_TASK_ID", "set-but-no-nvml")

	if r := ResolveGpuRuntime(); r != GpuRuntimeNone {
		t.Fatalf("no NVML should yield NONE; got %v", r)
	}
}

func TestResolveGpuRuntimeReturnsNoneWhenZeroDevices(t *testing.T) {
	withNVMLMock(t, 0)
	t.Setenv("MODAL_TASK_ID", "x")
	if r := ResolveGpuRuntime(); r != GpuRuntimeNone {
		t.Fatalf("0 devices should yield NONE; got %v", r)
	}
}

func TestResolveGpuRuntimeServerlessEnvVarsWinFirst(t *testing.T) {
	withNVMLMock(t, 1)
	// Set ALL serverless env vars at once; Modal takes priority.
	t.Setenv("MODAL_TASK_ID", "m1")
	t.Setenv("RUNPOD_POD_ID", "r1")
	t.Setenv("REPLICATE_MODEL", "rep1")
	if r := ResolveGpuRuntime(); r != GpuRuntimeModal {
		t.Fatalf("Modal env var should win; got %v", r)
	}

	os.Unsetenv("MODAL_TASK_ID")
	os.Unsetenv("MODAL_IMAGE_ID")
	if r := ResolveGpuRuntime(); r != GpuRuntimeRunpod {
		t.Fatalf("RunPod should win without Modal; got %v", r)
	}

	os.Unsetenv("RUNPOD_POD_ID")
	if r := ResolveGpuRuntime(); r != GpuRuntimeReplicate {
		t.Fatalf("Replicate should win as last serverless option; got %v", r)
	}
}

func TestResolveGpuRuntimeLambdaLabsByCloudDetect(t *testing.T) {
	withNVMLMock(t, 1)
	os.Unsetenv("MODAL_TASK_ID")
	os.Unsetenv("MODAL_IMAGE_ID")
	os.Unsetenv("RUNPOD_POD_ID")
	os.Unsetenv("RUNPOD_POD_HOSTNAME")
	os.Unsetenv("REPLICATE_MODEL")
	os.Unsetenv("REPLICATE_PREDICTION_ID")
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "lambda_labs"})
	t.Cleanup(cloud.ResetForTests)
	if r := ResolveGpuRuntime(); r != GpuRuntimeLambdaLabs {
		t.Fatalf("lambda_labs cloud should yield LAMBDA_LABS; got %v", r)
	}
}

func TestResolveGpuRuntimeCoreweaveByCloudDetect(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "coreweave"})
	t.Cleanup(cloud.ResetForTests)
	if r := ResolveGpuRuntime(); r != GpuRuntimeCoreweave {
		t.Fatalf("coreweave cloud should yield COREWEAVE; got %v", r)
	}
}

func TestResolveGpuRuntimeAWSEC2GPUByInstanceFamily(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	families := []string{"g4dn.xlarge", "g5.xlarge", "g6e.4xlarge", "p3.2xlarge", "p4d.24xlarge", "p4de.24xlarge", "p5.48xlarge", "p5e.48xlarge", "p5en.48xlarge"}
	for _, instance := range families {
		cloud.SetResultForTests(cloud.CloudEnv{Provider: "aws", Region: "us-east-1", InstanceType: instance})
		t.Cleanup(cloud.ResetForTests)
		if r := ResolveGpuRuntime(); r != GpuRuntimeAWSEC2GPU {
			t.Errorf("%s should yield AWS_EC2_GPU; got %v", instance, r)
		}
	}
}

func TestResolveGpuRuntimeGCPBundledFamilies(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	for _, instance := range []string{"a2-highgpu-1g", "a3-megagpu-8g", "g2-standard-4"} {
		cloud.SetResultForTests(cloud.CloudEnv{Provider: "gcp", Region: "us-central1", InstanceType: instance})
		t.Cleanup(cloud.ResetForTests)
		if r := ResolveGpuRuntime(); r != GpuRuntimeGCPGCEBundled {
			t.Errorf("%s should yield GCP_GCE_BUNDLED; got %v", instance, r)
		}
	}
}

func TestResolveGpuRuntimeGCPN1AttachedDecision9(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "gcp", Region: "us-central1", InstanceType: "n1-standard-8"})
	t.Cleanup(cloud.ResetForTests)
	if r := ResolveGpuRuntime(); r != GpuRuntimeGCPGCEN1Attached {
		t.Fatalf("N1 + NVML should yield GCP_GCE_N1_ATTACHED (Decision #9); got %v", r)
	}
}

func TestResolveGpuRuntimeAzureFamilies(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	for _, instance := range []string{"Standard_ND96isr_H100_v5", "Standard_NC24ads_A100_v4"} {
		cloud.SetResultForTests(cloud.CloudEnv{Provider: "azure", Region: "eastus", InstanceType: instance})
		t.Cleanup(cloud.ResetForTests)
		if r := ResolveGpuRuntime(); r != GpuRuntimeAzureVMGPU {
			t.Errorf("%s should yield AZURE_VM_GPU; got %v", instance, r)
		}
	}
}

func TestResolveGpuRuntimeAzureVGPUDecision10(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "azure", Region: "eastus", InstanceType: "Standard_NV6ads_A10_v5"})
	t.Cleanup(cloud.ResetForTests)
	if r := ResolveGpuRuntime(); r != GpuRuntimeAzureVMVGPU {
		t.Fatalf("NVadsA10_v5 should yield AZURE_VM_VGPU (Decision #10); got %v", r)
	}
}

func TestResolveGpuRuntimeUnknownCloudFallsToNone(t *testing.T) {
	withNVMLMock(t, 1)
	for _, k := range []string{"MODAL_TASK_ID", "MODAL_IMAGE_ID", "RUNPOD_POD_ID", "RUNPOD_POD_HOSTNAME", "REPLICATE_MODEL", "REPLICATE_PREDICTION_ID"} {
		os.Unsetenv(k)
	}
	cloud.SetResultForTests(cloud.CloudEnv{Provider: "aws", Region: "us-east-1", InstanceType: "t3.micro"})
	t.Cleanup(cloud.ResetForTests)
	if r := ResolveGpuRuntime(); r != GpuRuntimeNone {
		t.Fatalf("non-GPU AWS instance should yield NONE; got %v", r)
	}
}
